client/src/main/java/org/eclipse/hono/client/impl/DestinationCommandConsumer.java

```diff
@@ -0,0 +1,318 @@
+        final String originalDeviceId = msg.getAddress() != null
+                ? ResourceIdentifier.fromString(msg.getAddress()).getResourceId()
+                : null;
+        // look for a handler with the original device id first
+        final CommandHandlerWrapper commandHandler = getCommandHandlerOrDefault(originalDeviceId);
+        if (commandHandler != null) {
+            final Command command = Command.from(msg, tenantId, gatewayOrDeviceId);
+            final Tracer tracer = connection.getTracer();
+            // try to extract Span context from incoming message
+            final SpanContext spanContext = TracingHelper.extractSpanContext(tracer, msg);
+            final Span currentSpan = createSpan("send command", tenantId, gatewayOrDeviceId,
+                    tracer, spanContext);
+            logReceivedCommandToSpan(command, currentSpan);
+            commandHandler.getCommandHandler().handle(CommandContext.from(command, delivery, this.receiver, currentSpan));
+        } else {
+            LOG.error("no command handler found for command with device id {}, message address device id {} [tenant-id: {}]",
+                    gatewayOrDeviceId, originalDeviceId, tenantId);
+            ProtonHelper.released(delivery, true);
+        }
+    }
+
+    /**
+     * Checks whether a handler exists for the given device id.
+     * 
+     * @param deviceId The device id.
+     * @return {@code true} if a handler exists.
+     */
+    public boolean containsCommandHandler(final String deviceId) {
+        return commandHandlers.containsKey(deviceId);
+    }
+
+    /**
+     * Gets a handler for either the given device id or for the device id of <em>this</em>
+     * command consumer.
+     * 
+     * @param gatewayManagedDeviceId The id of a device connected to a gateway. May be {@code null}.
+     *                               If a handler for that device exists, it will be returned here.
+     *                               Otherwise the handler for <em>this</em> command consumers
+     *                               device id will be returned (if set).
+     * @return The handler or {@code null}.
+     */
+    public CommandHandlerWrapper getCommandHandlerOrDefault(final String gatewayManagedDeviceId) {
+        if (gatewayManagedDeviceId != null && !gatewayManagedDeviceId.equals(gatewayOrDeviceId)) {
+            final CommandHandlerWrapper handler = commandHandlers.get(gatewayManagedDeviceId);
+            if (handler != null) {
+                LOG.trace("using device specific command handler [gateway-id: {}, device-id: {}]", gatewayOrDeviceId,
+                        gatewayManagedDeviceId);
+                return handler;
+            }
+        }
+        return commandHandlers.get(gatewayOrDeviceId);
+    }
+
+    /**
+     * Gets the contained command handlers.
+     *
+     * @return The command handlers.
+     */
+    public Collection<CommandHandlerWrapper> getCommandHandlers() {
```

Comment:
This is needed for when a command consumer needs to be recreated because the existing consumer has been remotely closed. The new command consumer needs the same handlers as the old one.
This is part of the `CommandConsumerFactoryImpl` liveness check - see `CommandConsumerFactoryImpl.LivenessCheckData`.

I thought of adding a 
````
DestinationCommandConsumer.adoptHandlers(final DestinationCommandConsumer other)
````
method for that instead, but I found it better to have `Supplier<Collection<CommandHandlerWrapper>> commandHandlersSupplier` in the `LivenessCheckData`, instead of the `DestinationCommandConsumer` object.

Reply:
ok, maybe we can refactor that later ...